packages/react-dom/src/shared/ReactDOMUnknownPropertyHook.js

```diff
@@ -183,10 +183,12 @@ if (__DEV__) {
     if (typeof value === 'boolean') {
       warning(
         DOMProperty.shouldAttributeAcceptBooleanValue(name),
-        'Received `%s` for non-boolean attribute `%s`. If this is expected, cast ' +
-          'the value to a string.%s',
+        'If you intentionally tried to pass a boolean, pass it as a string ' +
+          'instead: `%s`="`%s`". If you mean to conditionally pass an ' +
+          'attribute, use a ternary expression: `%s`={condition ? value : null}.%s',
```

Comment:
If the value is `true` then this is very unlikely to happen because the problematic pattern is `condition && realValue` specifically. I can't think of a way where you'd end up with `true` for the same reason. Maybe we should have two separate error messages for `true` and `false` where this recommendation is only for the `false` case?

Maybe also call out that pattern specifically, so that people know what's wrong and what to look for.

Also, the recommendation should be to pass `undefined`, not `null`. It doesn't really matter for DOM components, but if it is a custom component that is the difference between defaultProps working and not working. So it's better to have consistent advice.